silkworm/node/db/access_layer.cpp

```diff
@@ -882,121 +874,261 @@ void write_last_finalized_block(RWTxn& txn, const evmc::bytes32& hash) {
+    std::vector<BlockHeader> sibling_headers;
+
+    // Read all siblings headers at specified height from db
+    process_headers_at_height(txn_, block_number, [&](BlockHeader&& header) {
+        sibling_headers.push_back(std::move(header));
+    });
+
+    // Read block header at specified height from snapshot (if any) just in case
+    std::optional<BlockHeader> header = read_header_from_snapshot(block_number);
+    if (header) {
+        sibling_headers.push_back(std::move(*header));
+    }
+
+    return sibling_headers;
 }
 
-bool DataModel::read_body(BlockNum block_number, const uint8_t (&hash)[kHashLength],
-                          bool read_senders, BlockBody& out) {
-    return db::read_body(txn_, block_number, hash, read_senders, out);
+bool DataModel::read_body(BlockNum height, HashAsArray hash, bool read_senders, BlockBody& body) const {
+    // Assume recent blocks are more probable: first lookup the block body in the db
+    const bool found = db::read_body(txn_, height, hash, read_senders, body);
+    if (found) return found;
+
+    return read_body_from_snapshot(height, read_senders, body);
 }
 
-bool DataModel::read_body(const evmc::bytes32& hash, BlockNum bn, BlockBody& body) {
-    return db::read_body(txn_, hash, bn, body);
+bool DataModel::read_body(const Hash& hash, BlockNum height, BlockBody& body) const {
+    return read_body(height, hash.bytes, /*read_senders=*/false, body);
+}
+
+bool DataModel::read_body(const Hash& hash, BlockBody& body) const {
+    const bool found = db::read_body(txn_, hash, body);
+    if (found) return found;
+
+    // Then search for it in the snapshots (if any)
+    const auto block_header{read_header_from_snapshot(hash)};
+    if (block_header) {
+        return read_body(block_header->number, hash.bytes, /*read_senders=*/false, body);
+    }
+
+    return false;
 }
 
-bool DataModel::read_body(const evmc::bytes32& hash, BlockBody& body) {
-    return db::read_body(txn_, hash, body);
+std::optional<evmc::bytes32> DataModel::read_canonical_hash(BlockNum height) const {
+    // Canonical chain is likely on db: first lookup there
+    auto canonical_hash{db::read_canonical_hash(txn_, height)};
+    if (canonical_hash) return canonical_hash;
+
+    // Then search for it in the snapshots (if any)
+    if (const auto block_header{read_header_from_snapshot(height)}; block_header) {
+        canonical_hash = block_header->hash();
+    }
+    return canonical_hash;
 }
 
-bool DataModel::read_canonical_block(BlockNum height, Block& block) {
-    return db::read_canonical_block(txn_, height, block);
+std::optional<BlockHeader> DataModel::read_canonical_header(BlockNum height) const {
+    // Canonical chain is likely on db: first lookup there
+    auto canonical_header{db::read_canonical_header(txn_, height)};
+    if (canonical_header) return canonical_header;
+
+    // Then search for it in the snapshots (if any)
+    return read_header_from_snapshot(height);
```

Comment:
Same as above. We need to implement it with a first read_canonical_hash(), then, if present we will read from db or snapshot.

```    
std::optional<evmc::bytes32> h = read_canonical_hash(txn, b);
if (!h) return std::nullopt;  // not found
return read_header(txn, b, h->bytes);  // we should raise exception if it fails
```